File: Roost/backend/src/main/java/com/travelplanner/roost/common/error/ErrorResponse.java
package com.travelplanner.roost.common.error;

import com.fasterxml.jackson.annotation.JsonInclude;
import java.util.Map;

public record ErrorResponse(ErrorBody error) {

    @JsonInclude(JsonInclude.Include.NON_NULL)
    public record ErrorBody(String code, String message, Map<String, String> details, String requestId) {}

    public static ErrorResponse of(
            ErrorCode code, String message, Map<String, String> details, String requestId) {
        Map<String, String> d = (details == null || details.isEmpty()) ? null : details;
        return new ErrorResponse(new ErrorBody(code.code(), message, d, requestId));
    }
}

File: TripWeaver/backend/Common/Errors/ErrorResponse.cs
using System.Text.Json.Serialization;

namespace TripWeaver.Common.Errors;

public sealed record ErrorResponse(ErrorBody Error)
{
    public static ErrorResponse Of(
        string code,
        string message,
        IReadOnlyDictionary<string, string>? details,
        string? requestId) =>
        new(new ErrorBody(code, message, details is { Count: > 0 } ? details : null, requestId));
}

public sealed record ErrorBody(
    string Code,
    string Message,
    [property: JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    IReadOnlyDictionary<string, string>? Details,
    string? RequestId);
